src/uri/mod.rs

```diff
@@ -813,28 +817,23 @@ fn parse_full(mut s: Bytes) -> Result<Uri, InvalidUri> {
     let authority_end = Authority::parse(&s[..])?;
 
     if scheme.is_none() {
+        // Path is not allowed if there is no scheme.
         if authority_end != s.len() {
             return Err(ErrorKind::InvalidFormat.into());
         }
-
-        let authority = Authority {
-            data: unsafe { ByteStr::from_utf8_unchecked(s) },
-        };
-
-        return Ok(Uri {
-            scheme: scheme.into(),
-            authority: authority,
-            path_and_query: PathAndQuery::empty(),
```

Comment:
This used to just assume `empty()`, and now its calling `parse(s)?`, right?

Reply:
Not exactly. It used to assume `PathAndQuery::empty()`  and now it is calling `PathAndQuery::from_shared(s)` where `s` is an empty `Bytes`. We know that `s` is an empty `Bytes` because we test that `authority_end` equals the whole length of `s` at line 821 (of the changed file) and then call `s.split_to(authority_end)` at line 831. This latter call becomes (in this case) `s.split_to(s.len())` which results in `s` being left empty.

`PathAndQuery::empty()` and `PathAndQuery::from_shared(s)` where `s` is empty both produce a `PathAndQuery` with `data` the equivalent of `ByteStr::new()` and `query` as `NONE` (the sentinel value declared in `uri::path`).

Logically this change still produces the same result in the `scheme.is_none()` case.

Performance wise, this particular case isn't covered by any of the existing benchmarks in `benches/uri.rs`. Although `PathAndQuery::from_shared(s)` in general is linear in the size of `s`, when `s` is empty this a constant operation, though this doesn't answer the performance question without having benchmarks.

If there are any concerns about the performance implications of this change I would be happy to add a benchmark to `benches/uri.rs` to cover this case.